ecommerce/api.py

```diff
@@ -286,3 +347,88 @@ def redeem_coupon(coupon_version, order):
         order=order, defaults={"coupon_version": coupon_version}
     )
     return coupon_redemption
+
+
+def get_new_order_by_reference_number(reference_number):
+    """
+    Parse a reference number received from CyberSource and lookup the corresponding Order.
+    Args:
+        reference_number (str):
+            A string which contains the order id and the instance which generated it
+    Returns:
+        Order:
+            An order
+    """
+    if not reference_number.startswith(_REFERENCE_NUMBER_PREFIX):
+        raise ParseException(
+            "Reference number must start with {}".format(_REFERENCE_NUMBER_PREFIX)
+        )
+    reference_number = reference_number[len(_REFERENCE_NUMBER_PREFIX) :]
+
+    try:
+        order_id_pos = reference_number.rindex("-")
+    except ValueError:
+        raise ParseException("Unable to find order number in reference number")
+
+    try:
+        order_id = int(reference_number[order_id_pos + 1 :])
+    except ValueError:
+        raise ParseException("Unable to parse order number")
+
+    prefix = reference_number[:order_id_pos]
+    if prefix != settings.CYBERSOURCE_REFERENCE_PREFIX:
+        log.error(
+            "CyberSource prefix doesn't match: %s != %s",
+            prefix,
+            settings.CYBERSOURCE_REFERENCE_PREFIX,
+        )
+        raise ParseException("CyberSource prefix doesn't match")
+
+    try:
+        return Order.objects.get(id=order_id)
+    except Order.DoesNotExist:
+        raise EcommerceException("Unable to find order {}".format(order_id))
+
+
+def enroll_user_on_success(order):  # pylint: disable=unused-argument
+    """
+    Check if the order is successful, then enroll the user
+
+    Args:
+        order (Order):
+            An order
+    """
+    # TBD: will be implemented later
+
+
+@transaction.atomic
+def create_unfulfilled_order(user):
+    """
+    Create a new Order which is not fulfilled for a purchasable course run. If course run is not purchasable,
+    it raises an Http404.
+
+    Args:
+        user (User):
+            The purchaser of the course run
+
+    Returns:
+        Order: A newly created Order for the CourseRun with the given course_id
+    """
+    # Note: validation is assumed to already have happen when the basket is being modified
+    basket, _ = Basket.objects.get_or_create(user=user)
+
+    order = Order.objects.create(status=Order.CREATED, purchaser=user)
+
+    for basket_item in basket.basketitems.all():
+        product_version = get_latest_product_version(basket_item.product)
+        Line.objects.create(
+            order=order, product_version=product_version, quantity=basket_item.quantity
```

Comment:
I think the idea is that you could buy seats in a course for other people. But I also agree that the REST API shouldn't need to handle that case since a user buying two seats for themself is not a use case we support.

Reply:
Ok I'll leave this as is, and the basket REST API will assume a quantity of 1 for all items in basket.

Reply:
We will need to allow bulk purchases via Cybersource, but that may require a separate API. Among other things, the fulfillment of a bulk purchase is to return a spreadsheet of 100% off coupon codes, and not an enrollment.